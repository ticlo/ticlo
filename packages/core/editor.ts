export * from './connect/ClientConn';
export * from './connect/ClientConnection';
export * from './block/Descriptor';
export * from './util/i18n';
export * from './util/Compare';
export {type DataMap, isDataTruncated, TRUNCATED} from './util/DataTypes';
export {encodeDisplay, decode, encode} from './util/Serialize';
export * from './util/Name';
export {endsWithNumberReg, getTailingNumber, smartStrCompare, isColorStr} from './util/String';
export * from './util/DateTime';
export {Logger} from './util/Logger';
export {stopPropagation, voidFunction} from './util/Functions';
export {PropDispatcher} from './block/Dispatcher';
export * from './util/Path';
