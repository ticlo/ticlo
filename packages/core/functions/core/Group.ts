import {BaseFunction, StatefulFunction} from '../../block/BlockFunction';
import {type BlockConfig} from '../../block/BlockProperty';
import {Functions} from '../../block/Functions';
import {type Block} from '../../block/Block';

export class GroupFunction extends BaseFunction<Block> {
  #autoUpdate = true;
  #setAutoUpdate(v: boolean) {
    if (v === this.#autoUpdate) {
      this.#autoUpdate = v;
      return true;
    }
    return false;
  }

  configChanged(config: BlockConfig, val: unknown): boolean {
    if (config._name === 'mode') {
      return this.#setAutoUpdate(config._value == null || config._value === 'auto');
    }
    return false;
  }

  run() {}

  cleanup(): void {}
}

Functions.add(GroupFunction, {
  name: 'group',
  icon: 'fas:folder',
  color: '9bd',
  properties: [],
});
