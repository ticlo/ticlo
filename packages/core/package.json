{
  "name": "@ticlo/core",
  "version": "0.0.11",
  "description": "ticlo core",
  "repository": {
    "type": "git",
    "url": "https://github.com/ticlo/ticlo.git"
  },
  "author": {
    "name": "Rick Zhou",
    "email": "rinick@gmail.com"
  },
  "license": "MPL-2.0",
  "main": "./index.ts",
  "module": "./index.ts",
  "types": "./index.ts",
  "dependencies": {
    "arrow-code": "^0.3.0",
    "axios": "1.7.4",
    "denque": "^2.1.0",
    "i18next": "^23.10.1",
    "lodash": "^4.17.21",
    "luxon": "^3.4.4",
    "qs": "^6.12.0"
  }
}
