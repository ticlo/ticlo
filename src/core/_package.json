{
  "name": "@ticlo/core",
  "version": "?",
  "description": "ticlo core",
  "repository": {
    "type": "git",
    "url": "https://github.com/ticlo/ticlo.git"
  },
  "author": {
    "name": "Rick Zhou",
    "email": "rinick@gmail.com"
  },
  "license": "MPL-2.0",
  "dependencies": {
    "arrow-code": "^0.3.0",
    "axios": "^1.8.1",
    "denque": "^2.1.0",
    "lodash": "^4.17.21",
    "luxon": "^3.5.0",
    "qs": "^6.14.0"
  }
}
