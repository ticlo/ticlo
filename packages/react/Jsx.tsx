import React from 'react';
import {JsFunction, SCRIPT_ERROR} from '@ticlo/core/functions/script/Js';
import {Functions} from '@ticlo/core/block/Functions';
import {Block} from '@ticlo/core/block/Block';
import {TicloComp} from './TicloComp';
import {validateReactComponent} from './validateReactComponent';
import {elementConfigs} from './BaseElement';
import {BlockIO, ErrorEvent} from '@ticlo/core';

const HOOKS = `const {
  useContext,
  useState,
  useEffect,
  useCallback,
  useMemo,
  useRef
} = React;`;

export class JsxFunction extends JsFunction {
  readonly _comp: React.ReactNode;
  _functionalComponent: Function;

  constructor(block: Block) {
    super(block);
    this._comp = <TicloComp key={block._blockId} block={block} />;
    this._data.output(this._comp);
  }
  inputChanged(input: BlockIO, val: any): boolean {
    // ignore parent implementation of inputChanged
    if (input._name === 'script') {
      this._compiledFunction = null;
      this._runFunction = null;
      this._functionalComponent = null;
      this._preProcessed = false;
      return val !== undefined;
    }
    return Boolean(this._runFunction || this._functionalComponent);
  }

  parseFunction(script: string): Function {
    const Babel = (window as any).Babel;

    if (Babel) {
      let toTransform = `"use strict";const React={};${HOOKS}function _F_(){${script}}`;
      let code = Babel.transform(toTransform, {presets: ['es2017', 'react']}).code;
      script = code.substring(33).replace('function _F_() ', '');
    }
    return new Function('React', script);
  }

  outputResult(result: unknown) {
    if (validateReactComponent(result)) {
      this._data.output(result, '#render');
      return this._comp;
    } else {
      this._data.output(undefined, '#render');
    }
    return result;
  }

  applyFunction(): any {
    let result: unknown;
    if (this._runFunction) {
      result = super.applyFunction();
    } else if (this._functionalComponent) {
      const Comp = this._functionalComponent;

      // @ts-ignore
      result = <Comp data={this.getDataProxy()} />;
    }
    return this.outputResult(result);
  }

  preProcessCompileResult() {
    let result: unknown;
    try {
      result = this._compiledFunction.call(this.getDataProxy(), React);
    } catch (err) {
      this._compiledFunction = null;
      return new ErrorEvent(SCRIPT_ERROR, err);
    }
    if (typeof result === 'function') {
      // let the function run again
      this._functionalComponent = result;
    } else {
      this._runFunction = this._compiledFunction;
      return this.outputResult(result);
    }
    return undefined;
  }
  cleanup() {
    this._data.output(undefined, '#render');
    super.cleanup();
  }
}

Functions.add(
  JsxFunction,
  {
    name: 'jsx',
    configs: elementConfigs,
    properties: [
      {
        name: 'script',
        type: 'string',
        mime: 'text/jsx',
        pinned: true,
      },
      {
        name: '#output',
        pinned: true,
        type: 'object',
        readonly: true,
      },
    ],
  },
  'react'
);
