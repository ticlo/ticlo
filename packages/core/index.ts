export {Block} from './block/Block';
export {Root, Flow, FlowFolder} from './block/Flow';
export {type FlowStorage, type Storage} from './block/Storage';
export {BlockProperty, BlockIO} from './block/BlockProperty';
export {BaseFunction, StatefulFunction, PureFunction} from './block/BlockFunction';
export {Functions} from './block/Functions';
export * from './block/FunctonData';
export * from './block/Event';
export * from './block/Descriptor';
export {ServerConnection} from './connect/ServerConnection';
export {type DataMap, TRUNCATED} from './util/DataTypes';
export {convertToObject} from './util/DataTruncate';
export {forAllPathsBetween} from './util/Path';
export {encodeDisplay, decode, encode, encodeSorted, decodeReviver} from './util/Serialize';
export * from './util/Name';
export {endsWithNumberReg, getTailingNumber, smartStrCompare, isColorStr} from './util/String';
export * from './util/DateTime';
export {Logger, addConsoleLogger} from './util/Logger';
export {stopPropagation, voidFunction} from './util/Functions';
export {PropDispatcher} from './block/Dispatcher';
export {resolvePath, getRelativePath, isBindable} from './util/Path';
export {setStorageFunctionProvider} from './functions/data/Storage';
export {setSecretCipher} from './block/Block';

// Export additional commonly used utilities
export * from './connect/ClientConnection';
export * from './connect/LocalConnection';
export {Uid} from './util/Uid';
export {escapedObject} from './util/EscapedObject';
export {arrowReplacer, arrowReviver} from './util/Serialize';
export {deepEqual, shallowEqual} from './util/Compare';
export {Resolver} from './block/Resolver';

// Export from block subdirectories
export * from './block/BlockFunction';
export * from './block/BlockProperty';
export * from './block/FunctonData';
export {StreamDispatcher} from './block/Dispatcher';

// Export utilities
export * from './util/Settings';
export * from './util/test-util';
export {scat} from './util/String';

// Export for web-server package
export * from './functions/web-server/HttpRequest';
export * from './functions/web-server/RouteFunction';

// Export common data functions
export * from './functions/data/CreateObject';
export * from './property-api/ObjectValue';

// register functions
import './functions/core/Group';
import './functions/Categories';
import './functions/math/Arithmetic';
import './functions/math/Compare';
import './functions/math/Boolean';
import './functions/string/CompareString';
import './functions/string/Join';
import './functions/string/Split';
import './functions/data/CreateObject';
import './functions/web-server/HttpRequest';
import './functions/web-server/RouteFunction';
import './functions/web-server/StaticResponse';
import './functions/http/CreateHeaders';
import './functions/http/HttpClient';
import './functions/http/Fetch';
import './functions/script/Js';
import './functions/data/State';
import './functions/date';
import './functions/time/Delay';
import './functions/time/Timer';
import './functions/condition/DefaultValue';
import './functions/condition/If';
import './worker/MapFunction';
import './worker/MultiWorkerFunction';
import './worker/HandlerFunction';
import './worker/WorkerFunction';
import './worker/SelectWorkerFunction';
