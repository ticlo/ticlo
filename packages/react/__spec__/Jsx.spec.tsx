import {Flow, Root} from '@ticlo/core';
import {transform} from '@babel/standalone';
import '../Jsx';

describe('Jsx', function () {
  (window as any).Babel = {transform};

  it('basic', async function () {
    let flow = new Flow();
    let aBlock = flow.createBlock('a');
    aBlock.setValue('#is', 'react:jsx');
    aBlock.setValue('script', 'return <span />');
    Root.run();
    expect(aBlock.getValue('#render')).toMatchSnapshot();
  });

  it('functional', async function () {
    let flow = new Flow();
    let aBlock = flow.createBlock('a');
    aBlock.setValue('#is', 'react:jsx');
    aBlock.setValue('value', 'v');
    aBlock.setValue(
      'script',
      `return function Foo({data}){
  return <span>{data.value}</span>;
  }`
    );
    Root.run();
    expect(aBlock.getValue('#render')).toMatchSnapshot();
  });
});
