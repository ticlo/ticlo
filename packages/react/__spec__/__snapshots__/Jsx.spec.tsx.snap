// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Jsx > basic 1`] = `<span />`;

exports[`Jsx > functional 1`] = `
<Foo
  data={
    {
      "script": "return function Foo({data}){
  return <span>{data.value}</span>;
  }",
      "value": "v",
    }
  }
/>
`;
